### PR TITLE
chore(expansion): add design proposal for LVM volume expansion

### PR DESCRIPTION
**Pull request by mittachaitu**
## Pull Request template

**Why is this PR required? What issue does it fix?**:
This PR adds LVM volume expansion workflow.

**What this PR does?**:
This PR adds a design proposal for LVM-Volume expansion.

**Does this PR require any upgrade changes?**:
No

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Any additional information for your reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [x ] Fixes #<issue number>
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated?
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

### CODE REVIEW

File: design/lvm/resize_workflow.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+it gets triggered upon updating PVC capacity.
+
+##### Controller Expansion
+
+- External-resizer is a sidecar in `openebs-lvm-controller` pod which watches for PVC capacity
+  updates. Once it receives the event external-resizer will make a `ControllerExpandVolume` gRPC request
+  to LVM-CSI controller plugin.
+- LVM-CSI controller plugin is another container in the same pod where external-resizer is running.
+  Once the controller plugin receives the `ControllerExpandVolume` gRPC request it does the following steps:
+  - List active snapshots on resizing volume, if there are any snapshots plugin will return an error
```
Comment (pawanpraka1) [inline design/lvm/resize_workflow.md:72]: can we explain why we are returning error here?

File: design/lvm/resize_workflow.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+Below is high level sequence diagram for volume expansion workflow
+
+![Volume Expansion Workflow](./images/resize_sequence_diagram.jpg)
+
+### Test Plan
+A test plan will include following test cases:
+- Test volume expansion operation on all supported filesystems(ext3, ext4, xfs, btrfs).
+- Test volume expansion while previous expansion of the volume is already in progress.
+- Restart of node LVM-Node-CSI-driver while filesystem expansion is in progress.
```
Comment (pawanpraka1) [inline design/lvm/resize_workflow.md:136]: test volume expansion with snapshot
test volume expansion for thick provisioned volume group (+ test case to over expand the volume that the vg szie)

Comment (mittachaitu) [reply design/lvm/resize_workflow.md:137]: updated

File: design/lvm/resize_workflow.md (hunk 1 above)
Comment (mittachaitu) [reply design/lvm/resize_workflow.md:72]: Updated the doc

File: design/lvm/resize_workflow.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+##### Controller Expansion
+
+- External-resizer is a sidecar in `openebs-lvm-controller` pod which watches for PVC capacity
+  updates. Once it receives the event external-resizer will make a `ControllerExpandVolume` gRPC request
+  to LVM-CSI controller plugin.
+- LVM-CSI controller plugin is another container in the same pod where external-resizer is running.
+  Once the controller plugin receives the `ControllerExpandVolume` gRPC request it does the following steps:
+  - List active snapshots on resizing volume, if there are any snapshots plugin will return an error as
+    a response to the request since LVM does not support online resizing of volume if snapshot(s) exist.
+  - If there are no snapshots then the plugin will update the capacity on the corresponding LVMVolume
```
Comment (avishnu) [inline design/lvm/resize_workflow.md:74]: Should this step be performed at the end after the filesystem-resize is successful?
If filesystem-resize is errored, will it go into reconcile mode till it is successful?

File: design/lvm/resize_workflow.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    resource and return success response to the request.
+- Once external-resizer gets success response it will mark PVC pending for filesystem expansion else
+  it will retry until it receives success response.
+
+Note: `ControllerExpandVolume` gRPC request retrun an error if volume has snapshots.
```
Comment (avishnu) [inline design/lvm/resize_workflow.md:79]: returns*

File: design/lvm/resize_workflow.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+## Proposal
+
+### User Stories
+
+As an application developer, I should be able to resize volume on the fly(When application consuming volume).
+
+Steps To Be Performed By User:
+1. Update the PVC capacity using `kubectl edit pvc <pvc_name>`
+
+Note: StorageClass should allow expanding the volume(i.e allowVolumeExpansion field must set to true).
```
Comment (avishnu) [inline design/lvm/resize_workflow.md:58]: must be*

File: design/lvm/resize_workflow.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+Note: `ControllerExpandVolume` gRPC request retrun an error if volume has snapshots.
+
+##### Filesystem Expansion
+
+- Container Orchestrator(CO)[kubelet] will send `NodeExpandVolume` gRPC request to LVM-CSI node driver
+  once it observes the PVC marked for filesystem expansion.
+- LVM-CSI node driver receives the `NodeExpandVolume` gRPC request and it will expand the lvm volume
+  using lvextend cli and trigger filesystem expansion. If expansion is successful it will return
+  success gRPC response.
```
Comment (avishnu) [inline design/lvm/resize_workflow.md:87]: If filesystem-resize is errored, will it go into reconcile mode till it is successful?

File: design/lvm/resize_workflow.md (hunk 3 above)
Comment (mittachaitu) [reply design/lvm/resize_workflow.md:74]: > Should this step be performed at the end after the filesystem-resize is successful?

During filesystem expansion request, node-driver will not receive desired capacity, In the LVM-LocalPV case, we are expanding volume as well as filesystem during filesystem expand call(i.e NodeExpandVolume). During `ControllerExpandVolume` request, we are updating [desired capacity](https://github.com/openebs/lvm-localpv/blob/22f412292d493f886ae5bc4925057753b934cead/pkg/apis/openebs.io/lvm/v1alpha1/lvmvolume.go#L72-L75) of the corresponding LVMVolume resource. 


> What will be the effect on the LVMVolume resource if filesystem-resize is errored?

Node-driver plugin will return an error to `NodeExapndVolume` request and Kubelet will send the request again and again(I think there should be some back-off time but not pretty sure about it) till it gets succeed.

File: design/lvm/resize_workflow.md (hunk 6 above)
Comment (mittachaitu) [reply design/lvm/resize_workflow.md:87]: Yes

**Review by avishnu — APPROVED** [review]
lgtm

**Review by pawanpraka1 — APPROVED** [review]
looks good.

**Pull request merged**